[Build] Build script for MacOS fails for targets older than 13.4 because tests can not be built

### Describe the issue

According to https://onnxruntime.ai/docs/build/inferencing.html, "By default, ONNX Runtime is configured to be built for a minimum target macOS version of 10.12." In practice, when the runtime is built without specifying a target OS it is built for the major OS version the build machine is running (15.0 when building on 15.4 in my case). When the `--apple_deploy_target` argument is set to anything less than 13.4, the build fails because the tests are using functionality that is not available on older MacOS versions. The runtime itself is compiled, but I couldn't find an option to disable building the tests.

### Urgency

_No response_

### Target platform

MacOS

### Build script

`./build.sh --cmake_generator Ninja --config Release --skip_submodule_sync --skip_tests --build_shared_lib --apple_deploy_target 10.15 --minimal_build --cmake_extra_defines CMAKE_OSX_ARCHITECTURES="x86_64;arm64" CMAKE_IGNORE_PATH=/opt/homebrew`

### Error / output

The first error is 

```
In file included from /Users/juhani.simola/work/onnxruntime-1.21.0/onnxruntime/test/common/logging/sinks_test.cc:3:
In file included from /Users/juhani.simola/work/onnxruntime-1.21.0/include/onnxruntime/core/common/common.h:23:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/chrono:969:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__chrono/formatter.h:25:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__chrono/ostream.h:33:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__format/format_functions.h:29:
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__format/formatter_floating_point.h:74:30: error: 'to_chars' is unavailable: introduced in macOS 13.4
  to_chars_result __r = std::to_chars(__first, __last, __value, __fmt, __precision);
```

### Visual Studio Version

_No response_

### GCC / Compiler Version

Xcode 15.4 and 16.3, SDK 16.3

Oh well. x86_64 build fails with Xcode 16.3 because of this too:
```
In file included from /Users/juhani.simola/work/onnxruntime-1.21.0/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx512vnni.cpp:26:
In file included from /Users/juhani.simola/work/onnxruntime-1.21.0/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx512_int8_blklen16.h:9:
In file included from /Users/juhani.simola/work/onnxruntime-1.21.0/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx512_int8_blklen32.h:8:
/Users/juhani.simola/work/onnxruntime-1.21.0/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:954:49: error: array index 4 is past the end of the array (that has type '__m256[4]') [-Werror,-Warray-bounds]
  954 |                 __m128 acc_1 = FoldAccumulators(acc[4], acc[5], acc[6], acc[7]);
```

--- a/tools/ci_build/github/apple/default_full_apple_framework_build_settings.json
+++ b/tools/ci_build/github/apple/default_full_apple_framework_build_settings.json
@@ -24,7 +24,7 @@
         ],
         "macosx": [
             "--macos=MacOSX",
-            "--apple_deploy_target=13.3"
+            "--apple_deploy_target=13.4"
         ],
         "iphoneos": [
             "--ios",

--- a/tools/ci_build/github/apple/default_training_ios_framework_build_settings.json
+++ b/tools/ci_build/github/apple/default_training_ios_framework_build_settings.json
@@ -33,7 +33,7 @@
         ],
         "macosx": [
             "--macos=MacOSX",
-            "--apple_deploy_target=13.3"
+            "--apple_deploy_target=13.4"
         ]
     }
 }

--- a/tools/ci_build/github/azure-pipelines/stages/py-cpu-packaging-stage.yml
+++ b/tools/ci_build/github/azure-pipelines/stages/py-cpu-packaging-stage.yml
@@ -222,7 +222,7 @@ stages:
           targetPath: $(Build.SourcesDirectory)/build/Release/dist/fixed_wheels
           artifactName: onnxruntime-macos-$(PythonVersion)
       variables:
-        MACOSX_DEPLOYMENT_TARGET: '13.3'
+        MACOSX_DEPLOYMENT_TARGET: '13.4'
       strategy:
         matrix:
           Python310:

--- a/tools/ci_build/github/azure-pipelines/templates/mac-cpu-packing-jobs.yml
+++ b/tools/ci_build/github/azure-pipelines/templates/mac-cpu-packing-jobs.yml
@@ -31,7 +31,7 @@ jobs:
   workspace:
     clean: all
   variables:
-    MACOSX_DEPLOYMENT_TARGET: '13.3'
+    MACOSX_DEPLOYMENT_TARGET: '13.4'
     ALLOW_RELEASED_ONNX_OPSET_ONLY: ${{ parameters.AllowReleasedOpsetOnly }}
     TODAY: $[format('{0:dd}{0:MM}{0:yyyy}', pipeline.startTime)]
     PROTO_CACHE_DIR: $(Pipeline.Workspace)/ccache_proto

